src/python/pants/engine/streaming_workunit_handler.py

```diff
@@ -116,7 +116,12 @@ def get_expanded_specs(self) -> ExpandedSpecs:
         for target in expanded_targets:
             source = targets_dict.get(str(target.address.spec), None)
             if source is None:
-                source = targets_dict[str(target.address.maybe_convert_to_target_generator())]
+                source = targets_dict.get(str(target.address.maybe_convert_to_target_generator()))
```

Comment:
Maybe the `targets_dict` could be initialized with both the `unexpanded_addresses` and all of the addresses we plan on looking up?

Or, if `target.address.spec` doesn't have a match, we could always initialize the `source` list to empty for `target.address.maybe_convert_to_target_generator()`, regardless of whether it was preinitialized?

Reply:
Basically, the bug appears to be that this code assumes that for a given set of `Specs`, the expanded `Targets` will always contain a subset of the unexpanded `Addresses`... and that doesn't appear to be the case for some inputs. So what you might want to do is log the `Specs` that are being used (which in your repro case appear to be generated)?

Reply:
The specs comes from `--changed-since=.. --changed-dependents=transitive`

Will add some logging to become aware when this happens and for what inputs, while injecting a new list ad-hoc seems like a workable approach too.